Support constants and functions for fluvio and some improvements to the derive macro

vrrashkov opened the pull request:
Hi, these are the new changes to the `fluvio-protocol-derive` and how it works. #3673

When passing values there are couple of ways you can do that: 

```rust
// Constant as a path
const TEST: i16 = 1;
#[fluvio(min_version = TEST)]

// Constant as a literal
const TEST: i16 = 1;
#[fluvio(min_version = "TEST")]

// Function as a literal
fn test() -> i16 { 1 }
#[fluvio(min_version = "test()")]

// Int
#[fluvio(min_version = 1)]
```

There are couple of improvements to how attributes are handled, which is with a new macro `parse_attributes!` this way is way more cleaner and will be much faster to add/update when needed.  I documented briefly some of the new functions/macros. 

Updated the enumerators and added min_version/max_version to it as previously they were not handled, again with the new macro. 

This PR also includes the changes from the 1st PR #3577. 

Here are some notes about couple of things i noticed:

The `ast/container.rs`: I have not updated the code here with the new macro, as I saw that currently those properties are not used anywhere in the project, so I kept it as is.

For the `ast/enum.rs`: Even though I added min_version/max_version, actually currently there is no logic behind that at all as previously it didn’t exist as well so not sure what should happen there, what kind of logic do we need to implement.

Connected to the above, because there is no logic for the enum variables, you can notice that inside [fluvio-spu-schema/server/smartmodule.rs](https://github.com/infinyon/fluvio/blob/0a4e03406db93c651fc3dbd1e0657a57281ab302/crates/fluvio-spu-schema/src/server/smartmodule.rs#L88) there are constants that are passed to `min_version/max_version` but they actually do not do anything as there is no logic for the enum properties. 

The [fluvio-spu-schema/produce/request.rs](https://github.com/infinyon/fluvio/blob/0a4e03406db93c651fc3dbd1e0657a57281ab302/crates/fluvio-spu-schema/src/produce/request.rs#L44C5-L44C5) has this field set with **min_version**:

```rust
#[fluvio(min_version = PRODUCER_TRANSFORMATION_API)]
pub smartmodules: Vec<SmartModuleInvocation>,
```

This is a non existent variable so in the 1st PR I changed this to use the real one, but because this was reverted and trough further research I noticed that this is actually not even used at all because we can see that in the file there is [Encoder/Decoder](https://github.com/infinyon/fluvio/blob/0a4e03406db93c651fc3dbd1e0657a57281ab302/crates/fluvio-spu-schema/src/produce/request.rs#L81) that are being implemented because of that it doesn’t even use the min_version in the proc-macro, so I decided to leave it like it is for now just so the changes are only for the `fluvio-protocol-derive` crate.

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,256 @@
+/// * `attr_ident` - The path ident to search for
+/// * `meta` - a return variable that will give us the parse_nested_meta of each element
+/// * `field` - nested path to search for
+/// * `opt` - mutable variable to save data into
```
The `mutable variable to save data into` comment contradicts with the only occurrence `if $opt.is_none()`.
Handling duplicates of parsed attributes should be part of `parse_attributes` operation.

vrrashkov replied:
Ah yes.. the option is actually just used as a way to check if the field is already used. Updated the docs.

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,256 @@
+}
+pub fn get_expr_value<'a>(
+    attr_name: &'a str,
+    field: &'a Option<Expr>,
```
Here and below `&'a Option<Expr>` should be replaced by `Option<&'a Expr>`. This will save from unnecessary cloning of `Exp`.

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,256 @@
+// We return $meta variable like shown here: https://github.com/rust-lang/rust-clippy/issues/1553
+pub fn parse_attributes_data(meta: &ParseNestedMeta) -> (Option<syn::Expr>, Span, String) {
+    // we can safely unwarp as this is already checked from the parse_attributes macro
+    let ident = meta.path.get_ident().unwrap();
```
Even if it's true for now, it might change later. This is a public function in `utils` crate, so `unwrap` and ignoring `Err` results for functions should not be introduced.

galibey replied:
The comment is still about the old version.

And, the result of `value.parse()` below is ignored.

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,256 @@
-                        }
+pub(crate) use parse_attributes;
+
+pub fn get_expr_value_from_meta(meta: &ParseNestedMeta) -> syn::Result<PropAttrsType> {
```
Misleading naming `get_expr_value_from_meta` but return type is `PropAttrsType`

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,256 @@
+}
 
-    None
+fn parse_int_expr<'a>(attr_name: &'a str, expr: &'a Expr, span: Span) -> syn::Result<i16> {
```
Can we make it generic over different ints `i32`, `i8`, etc.?

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,256 @@
+        ))
+    }
+}
+pub fn get_lit_int<'a>(
```
We have two functions here `get_lit_int` and `get_lit_int_value` that do the same thing. Can we make one from them that can cover both cases?

vrrashkov replied:
I kept them both but improved a bit on them. Also it was incorrect to call it `get_lit_int_value`. I renamed it to `get_expr_int_value` because even though we are returning Int value we need to check for the Unary operator as well and returns just the int value, that's also why I kept them both because the get_lit_int only does checks for int without the Unary and returns Lit.

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,256 @@
+) -> Option<MetaNameValue> {
+    for item in meta_list
+        .parse_args_with(Punctuated::<syn::Meta, Token![,]>::parse_terminated)
+        .unwrap_or_else(|err| panic!("Invalid #[new] attribute: {}", err))
```
We should use error with the proper span instead of panicking

vrrashkov replied:
I changed it so it returns **Result** instead of **Option** on most of the places and improved on the error handling.

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,256 @@
+        .cloned()
+}
+
+pub(crate) fn find_int_name_value(meta: &Meta, attr_name: &str) -> Option<u64> {
```
Misleading function name

galibey reviewed (changes requested):
Thanks again for your contribution @vrrashkov 

Apart from the comments below, since this is a new feature we need to also write unit-tests that cover the new functionality.

vrrashkov commented on crates/fluvio-protocol-derive/src/ast/enum.rs:
```diff
@@ -60,28 +64,31 @@ pub(crate) struct EnumProp {
     pub tag: Option<String>,
     pub discriminant: Option<DiscrimantExpr>,
     pub kind: FieldKind,
+    pub min_version: Option<PropAttrsType>,
```
I added them because I saw that here it's trying to use them but they were not existent. Should I just remove them even though this Enum is trying to use them?

https://github.com/infinyon/fluvio/blob/0a4e03406db93c651fc3dbd1e0657a57281ab302/crates/fluvio-spu-schema/src/server/smartmodule.rs#L88-L104

vrrashkov replied:
It won't change anything if I remove them, it will still work and still won't do anything of course :D

galibey replied:
Yes, so there are a couple of options here: 
1. make it work
2. leave as is and create a follow-up issue
3. remove wrong occurrences from the existing code, add compilation warning and create a follow-up ticket
Pick any.
Adding the unworking code to the unworking code is not a proper solution :)

vrrashkov replied:
I guess I lied a bit here :D Because of the new changes we are checking if attribute actually exists, so currently because there are some places where there are enums with min_version/max_version then if we remove it from here it throws similar to this error "`attribute min_version unrecognized`". So maybe lets leave it like that at the moment and with a new issue add the functionality for the enum versioning?

galibey replied:
Ok

vrrashkov commented on crates/fluvio-protocol-derive/src/ast/prop.rs:
```diff
@@ -151,76 +143,128 @@ impl UnnamedProp {
     }
 }
-
-pub fn validate_versions(min: i16, max: Option<i16>, field: Option<&str>) -> Option<String> {
```
This place was used to validate on compile time. Because min_version and max_version support constants and functions they can only be evaluated at run time, so we can't actually use the validate_versions as it was before, that's why I removed it. 

There is validation for the runtime which check the token stream directly. 

https://github.com/infinyon/fluvio/blob/4b9b32db623f39af18a0a8f95c75d66277d1e786/crates/fluvio-protocol-derive/src/ast/prop.rs#L47-L63

Even if we change that to throw errors with to_compile_error in the token stream it will not work again because the TokenStream that loads the values for min/max versions won't be evaluated.

galibey replied:
Ok, let's do it in runtime then. If users mess up with versions, at least they will receive a proper message during encoding/decoding, it's better than nothing.

vrrashkov replied:
We already have the runtime checks, so I haven't added anything more to it.

galibey replied:
What runtime check do you mean? We used to check that min is not greater than max, min is not negative, etc. We don't have it now, don't we?

vrrashkov replied:
There are checks here that check the version.

https://github.com/infinyon/fluvio/blob/4b9b32db623f39af18a0a8f95c75d66277d1e786/crates/fluvio-protocol-derive/src/ast/prop.rs#L47-L63

vrrashkov replied:
Previously the `validate_versions `only did checks and threw error of something was returned. We cannot throw syn::Error in this case. We need to have a code generation in our case in order to have the values evaluated before using them and then we throw an error. Not sure how you want the error to be thrown but currently the old code as I see prints the data with tracing is that not enough?

vrrashkov replied:
I made it so we make checks with the generated code and panic if it finds an issue. I used the original checks and errors as text. Does that look okay?

galibey replied:
I found a way of doing it in compile time for literals, consts and const fn. `assert` is stable in const context already. We can leverage it. See the example here https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a58bfe08d0bb0ea8f579e488d124b728

We can drop support for non-const function for now.

vrrashkov replied:
That is a pretty cool hack doing it like this: `const _: ()` . Currently I replaced all the panics with the const asserts and pretty much didn't change anything else, because even if you try to use a normal function it will still throw an error so no need to handle that differently than how it works.

vrrashkov commented:
@galibey  I added tests for the TokenStream how it should handle the different cases when passing values with the new type **PropAttrsType**. Let me know if you think more tests are needed for that functionality.

sehz commented:
Can you add to tests here?  https://github.com/infinyon/fluvio/blob/master/crates/fluvio-protocol/tests/version.rs

vrrashkov commented:
> Can you add to tests here? https://github.com/infinyon/fluvio/blob/master/crates/fluvio-protocol/tests/version.rs

Added couple of more cases covering the different types that can be passed and updated `test_encode_version`/`test_decode_version` accordingly. 

Also found an issue with the Unary operator so fixed that as well and updated the test.

galibey commented on crates/fluvio-protocol-derive/src/ast/prop.rs:
```diff
@@ -151,76 +143,322 @@ impl UnnamedProp {
+    Fn(Ident),
+    Int(i16),
+    #[default]
+    None,
```
Why do we need `None` variant? Property attributes can't be without type or with undefined type, right?

vrrashkov replied:
I thought it might be cleaner to do it like that so all the defaults are set in the `prop_attrs_type_value`, but it might be cleaner to do it without the `None` and just set the default as `Int(0)` so will change it with the new changes.

galibey commented on crates/fluvio-protocol-derive/src/ast/prop.rs:
```diff
@@ -151,76 +143,322 @@ impl UnnamedProp {
+///
+pub fn prop_attrs_type_value(
+    attrs_type: Option<&PropAttrsType>,
+    ident_type: Option<&Ident>,
```
This is always `None` except one occurrence which is in tests.

vrrashkov replied:
Currently yes but we might need this at any point in the future as currently the default is i16, but if there is even a single value that needs to be different than that, we would need this as I can see that there are places in the the proc-macro that are specifically set as `u8` even though they don't affect this specific case for now.

galibey replied:
In any case, this is hard to read. You are passing `ident_type: Option<&Ident>` which itself is only relevant for only one variant of `PropAttrsType`. 
And if there is no `attrs_type` (None), why do we return `0_i16`

vrrashkov replied:
I was just following how it worked previously that it defaults to 0 so I set it to return 0 always.

vrrashkov replied:
About how the type is parsed, let me know if you like this better. I added a new case IntParse, so if it is needed we can use that now instead of passing ident_type.

https://github.com/infinyon/fluvio/blob/f6d081c81a29c03d495aef8fde94a412f9c30f03/crates/fluvio-protocol-derive/src/ast/prop.rs#L164-L173

galibey replied:
Now clippy check fails. Just remove it for now because it is not used. We will add it later once it's needed.

vrrashkov replied:
Yep I saw that will delete for now but actually I was thinking we can even make it just Int(data, typ) without the IntParse, so we will always be parsing the type and this should be even better.

galibey commented on crates/fluvio-protocol-derive/src/ast/prop.rs:
```diff
@@ -151,76 +143,322 @@ impl UnnamedProp {
+    if let Some(attr) = attrs_type {
+        match &attr {
+            PropAttrsType::Lit(data) => parse_quote!(#data),
+            PropAttrsType::Fn(data) => TokenStream::from_str(&format!("{}()", data)).unwrap(),
```
```suggestion
            PropAttrsType::Fn(data) => parse_quote!(#data()),
```

galibey commented on crates/fluvio-protocol-derive/src/ast/prop.rs:
```diff
@@ -151,76 +143,322 @@ impl UnnamedProp {
+            PropAttrsType::Fn(data) => TokenStream::from_str(&format!("{}()", data)).unwrap(),
+            PropAttrsType::Int(data) => {
+                if let Some(itype) = ident_type {
+                    TokenStream::from_str(&format!("{}_{}", data, itype)).unwrap()
```
Need to find a way without a need for `unwrap` or properly propagate the error

vrrashkov replied:
I am not really sure if this is an issue to be unwrap here as this is just a TokenStream and if anything happens it will be in compile time. Maybe we can just use `expect` insteead of `unwrap`?

sehz replied:
https://docs.rs/syn/latest/syn/struct.Error.html

vrrashkov replied:
I don't think we can convert LexError with compile_error. From what I understand LexError already emits any errors but if we just want to remove the `unwrap` maybe we can do it like this.

`TokenStream::from_str(&format!("{}_i16", data)).map_err(|_| TokenStream::new())`

This way we don't have unwrap and still will be having all the errors.

galibey commented on crates/fluvio-protocol-derive/src/ast/prop.rs:
```diff
@@ -151,76 +143,322 @@ impl UnnamedProp {
     /// Sets this value to the field when it isn't present in the buffer.
     /// Example: `#[fluvio(default = "-1")]`
     pub default_value: Option<String>,
+    pub ignorable: Option<bool>,
```
If this is not used in code generation, we should not add it

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,273 @@
+    T: std::ops::Neg<Output = T> + FromStr,
+    <T as FromStr>::Err: std::fmt::Display,
+{
+    if let Some(value_expr) = value {
+        parse_int_expr::<T>(attr_name, value_expr, span)
+    } else {
+        Err(syn::Error::new(
+            span,
+            format!("Expected {attr_name} to be valid Expr: `{attr_name} = \"...\"`"),
+        ))
+    }
```
```suggestion
    let value_expr = value.ok_or_else(|| {
        syn::Error::new(
            span,
            format!("Expected {attr_name} to be valid Expr: `{attr_name} = \"...\"`"),
        )
    })?;
    parse_int_expr::<T>(attr_name, value_expr, span)
```

vrrashkov replied:
Removed the function because we don't need it anymore, because we have no more `Option<&Expr>` so no need for a separate function to unwrap, using directly the `parse_int_expr`.

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,273 @@
+}
+
+pub(crate) fn find_int_from_meta(meta: &Meta, attr_name: &str) -> syn::Result<u64> {
+    match find_expr_from_meta(meta, attr_name) {
+        Ok(ref value) => {
+            let value = get_lit_int(attr_name, Some(value), meta.span())?;
+
+            value.base10_parse::<u64>()
+        }
+        Err(err) => Err(err),
     }
```
```suggestion
    let value = find_expr_from_meta(meta, attr_name)?;
    let value = get_lit_int(attr_name, Some(&value), meta.span())?;
    value.base10_parse::<u64>()
```

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,273 @@
-            Lit::Str(val) => Some(val.clone()),
-            _ => unimplemented!(),
+pub(crate) fn find_string_from_meta<'a>(meta: &'a Meta, attr_name: &'a str) -> syn::Result<LitStr> {
+    match find_expr_from_meta(meta, attr_name) {
+        Ok(ref value) => {
+            let lit_str = get_lit_str(attr_name, Some(value), meta.span())?;
+            Ok(lit_str.clone())
```
```suggestion
    let value = find_expr_from_meta(meta, attr_name)?;
    let lit_str = get_lit_str(attr_name, Some(&value), meta.span())?;
    Ok(lit_str.clone())
```

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,273 @@
+    let (expr, attr_span, attr_name) = parse_attributes_data(meta)?;
+    get_expr_value(&attr_name, expr.as_ref(), attr_span)
+}
+pub fn get_expr_value<'a>(
```
Misleading naming: we are not getting expression value here, we detecting `PropAttrsType`

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,273 @@
+}
+pub fn get_expr_value<'a>(
+    attr_name: &'a str,
+    field: Option<&'a Expr>,
```
In almost all cases we pass here `Some()` and in only one case `None`. Maybe we could consider changing this value to `field: &'a Expr` and do the unwrapping in that one case, wdyt? This will make the code of this function more clear.

galibey commented on crates/fluvio-protocol-derive/src/util.rs:
```diff
@@ -1,75 +1,273 @@
+            // So to handle that we are checking if it's Unary Lit and continue as usual
+            // If needed this can be extended to handle the Unary operators
+            // But it doesn't seem that is necessary currently
+            Ok(PropAttrsType::Int(-parse_int_expr(attr_name, expr, span)?))
```
`parse_int_expr` is also handling `unary` expr inside. We should decide on which level we want to deal with it and do it only once.

vrrashkov replied:
They do both handle Unary, but in this because we are in the Unary case the type of `expr` is `Expr::Lit` so just using the `parse_int_expr` to get the Literal from there and convert it so we save some lines of code and set it as a negative because we know this is Unary. Of course in the future if needed we can use the other operators as well and the logic might change but i don't see a reason to do it for now as the only place Unary is used if -1 is being passed.

galibey commented on crates/fluvio-protocol-derive/src/ast/prop.rs:
````diff
@@ -151,76 +143,324 @@ impl UnnamedProp {
+/// #[fluvio(min_version = 1)]
+/// ```
+///
+/// None has a default Int value of 0 which is set in prop_attrs_type_value
````
```suggestion

```

galibey commented on crates/fluvio-protocol-derive/src/ast/prop.rs:
```diff
@@ -151,76 +143,324 @@ impl UnnamedProp {
+    Fn(Ident),
+    Int(i16),
+}
+impl Default for PropAttrsType {
```
Why do we need to impl `Default` for attr type? If the type is not identifiable, we should just throw an error.

vrrashkov replied:
I was just following how it worked previously that it defaults to 0 so decided to set a default value as well.

galibey replied:
If you mean that previously default value of `min_version: i16` is 0, that's fine. The equivalent of that for new `min_version: Option<PropAttrsType>` is None. This doesn't mean that the default needs for `PropAttrsType ` type.

vrrashkov replied:
I removed the default with the new changes.

galibey commented on crates/fluvio-protocol-derive/src/ast/prop.rs:
```diff
@@ -16,6 +21,38 @@ pub(crate) struct UnnamedProp {
+    let min = prop_attrs_type_value(min_prop);
+    let max = prop_attrs_type_value(max_props);
+
+    match (max_props, field) {
```
I don't understand the conditions here: we check `#min>#max` twice, we check `#min < 0` only if max_props is Some(_), why? We use `concat!` but don't do actual concatenation.

galibey commented on crates/fluvio-protocol-derive/Cargo.toml:
```diff
@@ -13,12 +13,8 @@ proc-macro = true
-features = ["full"]
-
-
+anyhow = { workspace = true }
```
We use `anyhow` only once in tests, and the usage is no different from `core::result::Result`, so we can drop it.
